src/csharp/Grpc.Core/Internal/MetadataArraySafeHandle.cs

```diff
@@ -66,16 +65,50 @@ public static Metadata ReadMetadataFromPtrUnsafe(IntPtr metadataArray)
                 var index = new UIntPtr(i);
                 UIntPtr keyLen;
                 IntPtr keyPtr = Native.grpcsharp_metadata_array_get_key(metadataArray, index, out keyLen);
-                string key = Marshal.PtrToStringAnsi(keyPtr, (int)keyLen.ToUInt32());
+                int keyLen32 = checked((int)keyLen.ToUInt32());
+                string key = WellKnownStrings.TryIdentify((byte*)keyPtr.ToPointer(), keyLen32)
+                    ?? Marshal.PtrToStringAnsi(keyPtr, keyLen32);
                 UIntPtr valueLen;
                 IntPtr valuePtr = Native.grpcsharp_metadata_array_get_value(metadataArray, index, out valueLen);
-                var bytes = new byte[valueLen.ToUInt64()];
-                Marshal.Copy(valuePtr, bytes, 0, bytes.Length);
-                metadata.Add(Metadata.Entry.CreateUnsafe(key, bytes));
+                int len32 = checked((int)valueLen.ToUInt64());
+                metadata.Add(Metadata.Entry.CreateUnsafe(key, (byte*)valuePtr.ToPointer(), len32));
             }
             return metadata;
         }
 
+        private static class WellKnownStrings
```

Comment:
this is looking good, but I think it would be better to make WellKnownString a top-level class under Grpc.Core.Internal and provide unit tests for it (that checks that well known strings are identified correctly and that other strings are not recognized as well-known by mistake).
I think it's fine to leave as is in this PR, provided that there'll be a followup PR that makes the refactoring and adds tests. WDYT?

Reply:
done and done